ADR/0039-workspace-deprecation.md

```diff
@@ -0,0 +1,98 @@
+
+## Decision
+
+1. We will remove the Workspace abstraction from Konflux and use standard Kubernetes namespaces
+directly for providing a place for users to run their pipelines and store their configurations
+and secrets.
+
+2. We will stop using the term Workspace, and start to use the term Namespace.
+
+3. The Konflux UI will expose a wizard for creating a new namespace. This
+wizard will be available to any user that has permissions to create namespaces.
+
+4. Konflux will provide an optional automation for creating a namespace for a user
+when he/she logins for the first time.
+
+5. Konflux won't be opinionated about the mechanism for initializing/maintaining namespaces with
+supporting resources such as ResourceQuota and LimitRange. Konflux will differ
+this responsibility to other tools which are specialized in this task such as 
+(but not limited to) [Kyverno](https://kyverno.io/policies/best-practices/add-ns-quota/add-ns-quota/) and [ArgoCD](https://github.com/konflux-ci/namespace-generator).
+
+6. Konflux will provide a thin [backend service](https://github.com/konflux-ci/workspace-manager) for listing the namespaces where the user has at least view access
+to the Konflux CRDs. This list will be used by the namespace switcher in the UI.
+This is required since the Kubernetes API doesn't let the user to list a subset
+of namespace. The user gets permissions to list all namespace or none.
+
+7. Konflux will provide ClusterRoles that will grant permissions to the Konflux
+and Tekton CRDs. Those will be (aggregated)[https://github.com/konflux-ci/konflux-ci/issues/440] to the built-in Kubernetes roles (`view`, `edit`, `admin`).
+
+8. Public viewer access will be provided by assigning the `view` role (see above) to the
+`system:authenticated` built-in group that contains all the authenticated users.
+
+9. Same as Kubernetes, Konflux won't have a resource for representing a user. Instead,
+it will use external Identity providers.
+
+9. Konflux won't provide a way for creating a ephemeral namespaces. It will differ this
+task to another tool.
+
+10. The `join the waitlist` button will be removed from the Konflux UI.
+
+11. Konflux won't provide a transparent multi-cluster deployment.
+If required, existing open source projects that handle multi-cluster deployments
+should be explored and Konflux should integrate with them.
+
+## Consequences
+
+1. By removing the Workspace abstraction, and delegating namespace management tasks to other open source tools,
+the Konflux developers will have more attention for developing features related to to Konflux' core mission.
+
+2. Operating Konflux will become easier for users who already familiar with Kubernetes concepts.
+
+3. Konflux operators can use their preferred way to manage namespaces on their clusters.
+
+4. Konflux integrates with other open source cloud native tools, which can improve Konflux
+fit in the the cloud native apps ecosystem.
+
+5. Konflux doesn't have a dependency on [kubesaw].
```

Comment:
I'm not sure that replacing KubeSaw by a bunch of other tools will make Konflux team's life easier. As @dperaza4dustbit mentioned most likely you will have to introduce some abstraction on top of all those 3rd party tools to be able to integrate them successfully anyway.
Alternatively you, of course, could just say that all the namespace, quota, security/permissions management, life-cycle etc is up to the Konflux admins to deal with and they are responsible for using whatever they want but it's already possible. 
KubeSaw doesn't lock you. And different Konflux instances are free to use whatever they want for resource management.
KubeSaw just provides a single solution for most of the features a Konflux system would need for resource management.
The only real technical con of using KubeSaw I see in this document is that KubeSaw currently requires OpenShift (not vanilla Kube). But this is something we can and should discuss.

Reply:
I added this document with all the features and two more possible implementation of workspace we can have in place upstream while KubeSaw gets to vanilla support status https://docs.google.com/document/d/1rIWCfC1zmHfClVIMO0xIDyn4-gMh25jFUzH3-9lJD_Y/edit#heading=h.5it1e62pzuqb

Reply:
@gbenhaim Let's clarify what exactly you want to deprecate/remove. Is it Workspace API and the proxy, plus a few additional things built around that? Or do you want to remove KubeSaw on its own? From what I have read in this document, it more looks like that you are trying to fix some issues with the proxy, workspace API, and association with the user. You don't need to use that if you don't want to. It was decision of the Konflux team to rely on this functionality. When you take a look at DevSandbox, it doesn't use this functionality at all and it's powered by KubeSaw as well.

Everything what you wrote in the document is either already supported by KubeSaw or is a planned functionality to be implemented soon. The goal of KubeSaw is to be a highly configurable tool that can be used either in a very lightweight mode, or with extended functionality (which can be enabled or disabled on demand).

Reply:
Kubesaw is great but the way we see the future of Konflux it's not really needed. We want to have single cluster deployment in order to make things simpler. In addition, I support that Konflux won't be opinionated about the way namespace are created and provisioned with initial set of resources since different Konflux admins would have different opinions about how to do it.

Reply:
I've mentioned at several places that KubeSaw works in single-cluster environment without any problem and at no extra costs. So let's put the multi-cluster discussion aside.
I understand that you want to keep user and namespace provisioning up to the stakeholders, but you still need to use something in the RH Konflux instance. Let's say that customer A decides to create namespaces manually and customer B uses ArgoCD - you may have X number of ways of provisioning namespaces with X number of customers, but this doesn't have any effect on what you decide to use in your instance. You can recommend something to customers (KubeSaw, ArgoCD, kubectl, kyverno,...), but none of them has to follow your recommendation.

Anyway, I would love to understand why you actually think that KubeSaw is not needed. You think that you won't have to provision namespaces using templates (create a set of resources in these namespaces) and keep them up to date? You don't think that you could have some lifecycle either of namespaces or users? How do you want to solve sharing? Do you want users to create RoleBindings directly? There are many other questions. Answering all these questions is out of the scope of this ADR, this is for a long discussion.

I would rather focus on what you are trying to solve in this ADR:
1. What technical problems you want to solve by removing KubeSaw?
2. Is it really the whole KubeSaw that you want to remove, or rather only some parts of it?

And also:
What are the Konflux's technical requirements for the NaaS solution?

Reply:
> I understand that you want to keep user and namespace provisioning up to the stakeholders, but you still need to use something in the RH Konflux instance.

I completely agree with this. I think we should make a clear distinction between a Konflux deployment and the Konflux project.

As I understand, this ADR is not necessarily about changing the RH Konflux deployment (although it will likely have an impact). This ADR is about removing a dependency from the Konflux project. It is stating that namespace management, for example, is outside the scope of the Konflux project.

Of course, when deploying Konflux in your organization, then there are other factors to consider. Maybe kubesaw is the thing you need, maybe you already have this handled by your org's IT team, maybe ArgoCD meets all your requirements. From Konflux's perspective, it shouldn't matter.

Reply:
I would prefer to mention those tools in this ADR in order to show that there are alternatives that **can** be used.

Reply:
If the are no objections I'll close this thread.